### PR TITLE
Optimize Enum.slice/2 and Enum.slice/3

### PR DESCRIPTION
**Pull request by eksperimental**

### CODE REVIEW

**Comment by josevalim** [comment]
Hi @eksperimental! Do we have some measurements here? Because those gains seem to be very tiny and I am not sure it is worth the code complexity (we are adding 26LOC).

File: lib/elixir/lib/enum.ex [hunk 1]
```diff
@@ -2662,21 +2695,14 @@ defmodule Enum do
 
   ## slice
 
-  defp do_slice([], _start, _count) do
```
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: This is just changing the code style, isn't?

Comment (eksperimental) [inline lib/elixir/lib/enum.ex]: it was only line L#2666 and L#2670 that were untouched by this commit. the rest of the lines were modified so i decided to make it compact as it is now

**Comment by eksperimental** [comment]
I don't have measurements. I can take them but they will have to be against changes introduced in #4983. the whole idea is to delegate to Enum.fetch and take advantage of the new optimizations.
I think if we pass data that Enum.fetch/2 was optimized for (big maps and structs) and big ranges, the will be a difference, just before there is a difference in Enum.fetch/2 before and after.

the other optimization is to avoid pattern matching twice on a range, which showed me in the past to be more expensive than matching once and then match on the matched integers.

let me know if you still want the benchmarks, I can take care of that if needed.

**Comment by josevalim** [comment]
@eksperimental I believe we should measure the two changes in isolation. As said, we are increasing code size with the justification of performance, so we need to confirm it is true.

**Comment by eksperimental** [comment]
any micro-benchmarking tool that can you recommend for the job?

**Comment by josevalim** [comment]
Benchee or benchfella will likely be fine!

**Comment by josevalim** [comment]
@eksperimental I assume we can close this one?

**Comment by eksperimental** [comment]
yes we can,
I did the tests and it actually there were many improvement, but it showed me where we were still failing.
Therefore I ended up doing https://github.com/elixir-lang/elixir/pull/5017

**Pull request closed without merging**